File: windows/CaptureCopy.cs
namespace Snippr;

/// UI language for capture-flow copy. Missing or garbage → English (owner default).
public enum UILanguage
{
    English,
    Vietnamese,
}

public static class UILanguageUtil
{
    public const string EnglishCode = "en";
    public const string VietnameseCode = "vi";

    public static UILanguage Parse(string? raw) =>
        string.Equals(raw, VietnameseCode, StringComparison.OrdinalIgnoreCase)
            ? UILanguage.Vietnamese
            : UILanguage.English;

    public static string Code(UILanguage lang) =>
        lang == UILanguage.Vietnamese ? VietnameseCode : EnglishCode;

    /// Labels stay in their own language so the picker is findable either way.
    public static string MenuLabel(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Tiếng Việt",
        _ => "English",
    };
}

/// Capture-flow strings. Production reads only this table. Gates pin the
/// same sentences as *independent literals* so editing one cell goes red.
/// Keys match macOS `CaptureCopy.swift` after `3519422`; Windows uses Ctrl+C
/// and `px` where the platform differs. `Resolve` is bound by the app so
/// ParityGate can compile this file without `AppSettings`. Language has no
/// default: omitting it is a compile error (Honey W1). Call sites pass
/// `Current` or a session snapshot.
public static class CaptureCopy
{
    public static Func<UILanguage> Resolve { get; set; } = static () => UILanguage.English;

    public static UILanguage Current => Resolve();

    public static string StopHint(bool hotkeysRegistered, UILanguage language) =>
        hotkeysRegistered ? StopHintHotkey(language) : StopHintFallback(language);

    public static string StopHintHotkey(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Enter/✓ xong · Ctrl+C copy · Esc hủy",
        _ => "Enter/✓ done · Ctrl+C copy · Esc cancel",
    };

    public static string StopHintFallback(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "bấm ✓ để xong",
        _ => "click ✓ to finish",
    };

    public static string StitchingProgress(
        int pixels, bool hotkeysRegistered, UILanguage language)
    {
        var stitched = language == UILanguage.Vietnamese
            ? $"Đã ghép {pixels} px"
            : $"Stitched {pixels} px";
        var keep = language == UILanguage.Vietnamese ? " — cuộn tiếp · " : " — keep scrolling · ";
        return stitched + keep + StopHint(hotkeysRegistered, language);
    }

    public static string StitchingProgressUp(
        int pixels, bool hotkeysRegistered, UILanguage language)
    {
        var stitched = language == UILanguage.Vietnamese
            ? $"Đã ghép {pixels} px"
            : $"Stitched {pixels} px";
        var up = language == UILanguage.Vietnamese ? " (nối lên trên)" : " (connecting upward)";
        var keep = language == UILanguage.Vietnamese ? " — cuộn tiếp · " : " — keep scrolling · ";
        return stitched + up + keep + StopHint(hotkeysRegistered, language);
    }

    public static string DoneButton(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "✓ Xong",
        _ => "✓ Done",
    };

    public static string ScrollCancelled(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Đã hủy chụp cuộn",
        _ => "Scrolling capture cancelled",
    };

    public static string Copied(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Đã copy",
        _ => "Copied",
    };

    public static string RegionTooSmall(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Vùng quá nhỏ cho chụp cuộn — chọn vùng cao hơn 60 px",
        _ => "Region too small for scrolling capture — pick an area taller than 60 px",
    };

    public static string Bidirectional(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Cuộn lên hoặc xuống — stitcher nối cả hai chiều",
        _ => "Scroll up or down — stitching works both ways",
    };

    public static string NoMatch(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Chưa khớp được — cuộn chậm lại một chút",
        _ => "No match — scroll a bit slower",
    };

    public static string WaitingFirstFrame(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Chờ khung hình đầu tiên…",
        _ => "Waiting for the first frame…",
    };

    public static string CompatibilityMode(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Đang dùng chế độ tương thích — cuộn tiếp · ",
        _ => "Using compatibility mode — keep scrolling · ",
    };

    public static string Retrace(int pixels, UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => $"Đang cuộn qua vùng đã chụp — {pixels} px",
        _ => $"Scrolling through captured area — {pixels} px",
    };

    public static string LostSegment(int n, UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese =>
            $"Mất một đoạn do cuộn quá nhanh — đang ghi đoạn {n}; vạch sáng đánh dấu chỗ thiếu",
        _ =>
            $"Missed a stretch from scrolling too fast — recording segment {n}; a bright bar marks the gap",
    };

    public static string BackendSync(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Đang đồng bộ chế độ tương thích — cuộn chậm để nối tiếp",
        _ => "Syncing compatibility mode — scroll slowly to reconnect",
    };

    public static string BackendNewSegment(int n, UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese =>
            $"Đã đổi chế độ chụp và bắt đầu đoạn {n}; vạch sáng đánh dấu chỗ thiếu",
        _ =>
            $"Capture mode changed and started segment {n}; a bright bar marks the gap",
    };

    public static string ScreenshotCopied(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Đã copy",
        _ => "Screenshot copied",
    };

    public static string ScreenshotCopiedToClipboard(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Đã copy vào clipboard",
        _ => "Screenshot copied to clipboard",
    };

    public static string ScreenshotSummary(IReadOnlyList<string> parts, UILanguage lang)
    {
        var joined = string.Join(" · ", parts);
        return lang == UILanguage.Vietnamese ? $"Ảnh {joined}" : $"Screenshot {joined}";
    }

    public static string ToastCopiedWord(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "đã copy",
        _ => "copied",
    };

    public static string ToastSavedWord(string filename, UILanguage lang) =>
        lang == UILanguage.Vietnamese ? $"đã lưu {filename}" : $"saved {filename}";

    public static string ToastSaveFailed(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "lưu thất bại",
        _ => "save failed",
    };

    public static string ToastSaveFailedCopied(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "lưu thất bại — đã copy vào clipboard",
        _ => "save failed — copied instead",
    };

    /// Sentence-initial save toast (review `.save`), not the "saved x.png" fragment.
    public static string SavedFile(string filename, UILanguage lang) =>
        lang == UILanguage.Vietnamese ? $"Đã lưu {filename}" : $"Saved {filename}";

    public static string SaveFailed(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Lưu thất bại",
        _ => "Save failed",
    };

    public static string ExportAnnotatedFailed(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Không xuất được ảnh có nét vẽ — thử lại",
        _ => "Couldn't export the annotated image — try again",
    };

    public static string SavedRegionGone(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Vùng đã lưu không còn trên màn hình — chọn lại nhé",
        _ => "The saved region is no longer on screen — pick it again",
    };

    public static string SelectionTooLargeForBackdrop(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Vùng chọn quá lớn cho Backdrop",
        _ => "Selection too large for Backdrop",
    };

    public static string BackdropBuildFailed(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Không dựng được nền Backdrop — thử preset khác",
        _ => "Couldn't build the Backdrop — try another preset",
    };

    public static string NoPreviousArea(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Chưa có vùng đã lưu — chụp một vùng trước đã",
        _ => "No previous area — use Capture Area first",
    };

    public static string ScreenRecordingNeeded(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Cần quyền Screen Recording — bật Snippr trong Cài đặt hệ thống",
        _ => "Screen Recording permission needed — enable Snippr in System Settings",
    };

    public static string CaptureFailed(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Chụp thất bại",
        _ => "Capture failed",
    };

    public static string NoWindowFound(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Không tìm thấy cửa sổ",
        _ => "No window found",
    };

    /// Win-only: the live-preview prompt before the first stitched frame.
    /// macOS deleted the dead `scrollSlowly` path; Windows still shows it.
    public static string ScrollSlowly(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Cuộn từ từ — ảnh ghép hiện bên dưới",
        _ => "Scroll slowly — the stitch appears below",
    };

    /// Win-only: clipboard held by another process (RouteReviewed copy).
    public static string ClipboardBusy(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Clipboard đang bận — thử lại sau giây lát",
        _ => "Clipboard is busy — try again in a moment",
    };

    /// Win-only fragment inside the screenshot summary.
    public static string ClipboardBusyWord(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "clipboard bận",
        _ => "clipboard busy",
    };

    public static string ClipboardBusySaved(string filename, UILanguage lang) =>
        lang == UILanguage.Vietnamese
            ? $"Clipboard bận — đã lưu {filename}"
            : $"Clipboard busy — saved {filename}";

    public static string CopySaveFailedOpenEditor(UILanguage lang) => lang switch
    {
        UILanguage.Vietnamese => "Không copy/lưu được — mở editor",
        _ => "Couldn't copy or save — opening the editor",
    };
}

File: Sources/Snippr/CaptureCopy.swift
import Foundation

/// UI language for capture-flow copy. Missing or garbage → English (owner default).
enum UILanguage: String, CaseIterable {
    case english = "en"
    case vietnamese = "vi"

    static var resolved: UILanguage {
        let raw = UserDefaults.standard.string(forKey: Settings.Keys.uiLanguage) ?? "en"
        return UILanguage(rawValue: raw) ?? .english
    }

    /// Labels stay in their own language so the picker is findable either way.
    var menuLabel: String {
        switch self {
        case .english: return "English"
        case .vietnamese: return "Tiếng Việt"
        }
    }
}

/// Capture-flow strings. Production reads only this table. Gates pin the
/// same sentences as *independent literals* so editing one cell goes red.
/// Windows PR repeats these keys/literals in `CaptureCopy.cs`.
enum CaptureCopy {
    static func stopHint(hotkeysRegistered: Bool, language: UILanguage = .resolved) -> String {
        hotkeysRegistered ? stopHintHotkey(language) : stopHintFallback(language)
    }

    static func stopHintHotkey(_ lang: UILanguage) -> String {
        switch lang {
        case .english: return "Enter/✓ done · ⌘C copy · Esc cancel"
        case .vietnamese: return "Enter/✓ xong · ⌘C copy · Esc hủy"
        }
    }

    static func stopHintFallback(_ lang: UILanguage) -> String {
        switch lang {
        case .english: return "click ✓ to finish"
        case .vietnamese: return "bấm ✓ để xong"
        }
    }

    static func stitchingProgress(
        points: Int, connectingUp: Bool, hotkeysRegistered: Bool,
        language: UILanguage = .resolved
    ) -> String {
        let stitched: String
        switch language {
        case .english: stitched = "Stitched \(points) pt"
        case .vietnamese: stitched = "Đã ghép \(points) pt"
        }
        let up: String
        switch language {
        case .english: up = connectingUp ? " (connecting upward)" : ""
        case .vietnamese: up = connectingUp ? " (nối lên trên)" : ""
        }
        let keep: String
        switch language {
        case .english: keep = " — keep scrolling · "
        case .vietnamese: keep = " — cuộn tiếp · "
        }
        return stitched + up + keep + stopHint(hotkeysRegistered: hotkeysRegistered, language: language)
    }

    static func doneButton(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "✓ Done"
        case .vietnamese: return "✓ Xong"
        }
    }

    static func scrollCancelled(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Scrolling capture cancelled"
        case .vietnamese: return "Đã hủy chụp cuộn"
        }
    }

    static func copied(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Copied"
        case .vietnamese: return "Đã copy"
        }
    }

    static func regionTooSmall(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english:
            return "Region too small for scrolling capture — pick an area taller than 60 pt"
        case .vietnamese:
            return "Vùng quá nhỏ cho chụp cuộn — chọn vùng cao hơn 60 pt"
        }
    }

    static func bidirectional(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Scroll up or down — stitching works both ways"
        case .vietnamese: return "Cuộn lên hoặc xuống — stitcher nối cả hai chiều"
        }
    }

    static func noMatch(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "No match — scroll a bit slower"
        case .vietnamese: return "Chưa khớp được — cuộn chậm lại một chút"
        }
    }

    static func waitingFirstFrame(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Waiting for the first frame…"
        case .vietnamese: return "Chờ khung hình đầu tiên…"
        }
    }

    static func compatibilityMode(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Using compatibility mode — keep scrolling · "
        case .vietnamese: return "Đang dùng chế độ tương thích — cuộn tiếp · "
        }
    }

    static func retrace(_ points: Int, _ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Scrolling through captured area — \(points) pt"
        case .vietnamese: return "Đang cuộn qua vùng đã chụp — \(points) pt"
        }
    }

    static func lostSegment(_ n: Int, _ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english:
            return "Missed a stretch from scrolling too fast — recording segment \(n); a bright bar marks the gap"
        case .vietnamese:
            return "Mất một đoạn do cuộn quá nhanh — đang ghi đoạn \(n); vạch sáng đánh dấu chỗ thiếu"
        }
    }

    static func backendSync(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Syncing compatibility mode — scroll slowly to reconnect"
        case .vietnamese: return "Đang đồng bộ chế độ tương thích — cuộn chậm để nối tiếp"
        }
    }

    static func backendNewSegment(_ n: Int, _ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english:
            return "Capture mode changed and started segment \(n); a bright bar marks the gap"
        case .vietnamese:
            return "Đã đổi chế độ chụp và bắt đầu đoạn \(n); vạch sáng đánh dấu chỗ thiếu"
        }
    }

    static func screenshotCopied(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Screenshot copied"
        case .vietnamese: return "Đã copy"
        }
    }

    static func screenshotCopiedToClipboard(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Screenshot copied to clipboard"
        case .vietnamese: return "Đã copy vào clipboard"
        }
    }

    static func screenshotSummary(_ parts: [String], _ lang: UILanguage = .resolved) -> String {
        let joined = parts.joined(separator: " · ")
        switch lang {
        case .english: return "Screenshot \(joined)"
        case .vietnamese: return "Ảnh \(joined)"
        }
    }

    static func toastCopiedWord(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "copied"
        case .vietnamese: return "đã copy"
        }
    }

    static func toastSavedWord(_ filename: String, _ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "saved \(filename)"
        case .vietnamese: return "đã lưu \(filename)"
        }
    }

    static func toastSaveFailed(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "save failed"
        case .vietnamese: return "lưu thất bại"
        }
    }

    static func toastSaveFailedCopied(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "save failed — copied instead"
        case .vietnamese: return "lưu thất bại — đã copy vào clipboard"
        }
    }

    /// Sentence-initial save toast (router `.save`), not the "saved x.png" fragment.
    static func savedFile(_ filename: String, _ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Saved \(filename)"
        case .vietnamese: return "Đã lưu \(filename)"
        }
    }

    static func saveFailed(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Save failed"
        case .vietnamese: return "Lưu thất bại"
        }
    }

    static func exportAnnotatedFailed(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Couldn't export the annotated image — try again"
        case .vietnamese: return "Không xuất được ảnh có nét vẽ — thử lại"
        }
    }

    static func savedRegionGone(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "The saved region is no longer on screen — pick it again"
        case .vietnamese: return "Vùng đã lưu không còn trên màn hình — chọn lại nhé"
        }
    }

    static func selectionTooLargeForBackdrop(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Selection too large for Backdrop"
        case .vietnamese: return "Vùng chọn quá lớn cho Backdrop"
        }
    }

    static func backdropBuildFailed(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Couldn't build the Backdrop — try another preset"
        case .vietnamese: return "Không dựng được nền Backdrop — thử preset khác"
        }
    }

    static func noPreviousArea(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "No previous area — use Capture Area first"
        case .vietnamese: return "Chưa có vùng trước — hãy chọn vùng trước đã"
        }
    }

    static func screenRecordingNeeded(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english:
            return "Screen Recording permission needed — enable Snippr in System Settings"
        case .vietnamese:
            return "Cần quyền Screen Recording — bật Snippr trong Cài đặt Hệ thống"
        }
    }

    static func captureFailed(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "Capture failed"
        case .vietnamese: return "Chụp thất bại"
        }
    }

    static func noWindowFound(_ lang: UILanguage = .resolved) -> String {
        switch lang {
        case .english: return "No window found"
        case .vietnamese: return "Không tìm thấy cửa sổ"
        }
    }
}
